topology/nhlt/intel/dmic-nhlt.c

```diff
@@ -0,0 +1,1002 @@
+			       fifo_word_length, clk_min, clk_max, duty_min, duty_max, sample_rate,
+			       unmute_ramp_time_ms);
+}
+
+static int set_pdm_data(struct tplg_pre_processor *tplg_pp,
+			snd_config_t *cfg)
+{
+	snd_config_iterator_t i, next;
+	snd_config_t *class_cfg;
+	int mic_a_enable;
+	int mic_b_enable;
+	snd_config_t *n;
+	const char *id;
+	int polarity_a;
+	int polarity_b;
+	int clk_edge;
+	long int_val;
+	int ctrl_id;
+	int skew;
+	int ret;
+
+	/* get default values */
+	ret = snd_config_search(tplg_pp->input_cfg, "Class.Base.pdm_config", &class_cfg);
+	if (ret < 0)
+		return -EINVAL;
+
+	/* set default values */
+	snd_config_for_each(i, next, class_cfg) {
+
+		n = snd_config_iterator_entry(i);
+
+		if (snd_config_get_id(n, &id) < 0)
+			continue;
+
+		if (!strcmp(id, "ctrl_id")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			ctrl_id = int_val;
+		}
+
+		if (!strcmp(id, "mic_a_enable")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			mic_a_enable = int_val;
+		}
+
+		if (!strcmp(id, "mic_b_enable")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			mic_b_enable = int_val;
+		}
+
+		if (!strcmp(id, "polarity_a")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			polarity_a = int_val;
+		}
+
+		if (!strcmp(id, "polarity_b")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			polarity_b = int_val;
+		}
+
+		if (!strcmp(id, "clk_edge")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			clk_edge = int_val;
+		}
+
+		if (!strcmp(id, "skew")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			skew = int_val;
+		}
+	}
+
+	/* set object values */
+	snd_config_for_each(i, next, cfg) {
+
+		n = snd_config_iterator_entry(i);
+
+		if (snd_config_get_id(n, &id) < 0)
+			continue;
+
+		if (!strcmp(id, "ctrl_id")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			ctrl_id = int_val;
+		}
+
+		if (!strcmp(id, "mic_a_enable")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			mic_a_enable = int_val;
+		}
+
+		if (!strcmp(id, "mic_b_enable")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			mic_b_enable = int_val;
+		}
+
+		if (!strcmp(id, "polarity_a")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			polarity_a = int_val;
+		}
+
+		if (!strcmp(id, "polarity_b")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			polarity_b = int_val;
+		}
+
+		if (!strcmp(id, "clk_edge")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			clk_edge = int_val;
+		}
+
+		if (!strcmp(id, "skew")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			skew = int_val;
+		}
+	}
+
+	return dmic_set_pdm_params(tplg_pp, ctrl_id, mic_a_enable, mic_b_enable, polarity_a,
+				   polarity_b, clk_edge, skew);
+}
+
+static int set_mic_data(struct tplg_pre_processor *tplg_pp, snd_config_t *cfg)
```

Comment:
you call this a set but its really a get isnt it? you're getting the user values from the input conf file and updating your local stuct for processing later

Reply:
heh well, I guess it could be thought both ways. I was thinking it this way: "I'm setting the data to internal structs from topology" then "getting the calculated endpoints from the engine". So I see the nhlt part as an external engine where you need to set data, then ask it to generate the blobs